### PR TITLE
WS settings/set - Add validations and restrict to defined Settings

### PR DESCRIPTION
**Pull request by teryk**

### CODE REVIEW

File: sonar-plugin-api/src/main/java/org/sonar/api/config/PropertyDefinition.java [hunk 1]
```diff
@@ -150,41 +160,63 @@ static PropertyDefinition create(Property annotation) {
...
+      }
+    };
+  }
+
+  private static Validation validateRegexp() {
+    return value -> {
+      try {
+        Pattern.compile(value);
+        return Result.SUCCESS;
+      } catch (PatternSyntaxException e) {
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/config/PropertyDefinition.java]: MAJOR Either log or rethrow this exception. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1166)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* MAJOR 2 major

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [SetAction.java#L77](https://github.com/SonarSource/sonarqube/blob/e8294dc6eda237e4391ec29867028a6185e42931/server/sonar-server/src/main/java/org/sonar/server/setting/ws/SetAction.java#L77): Remove this unused "i18n" private field. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1068)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* MAJOR 2 major

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [SetAction.java#L77](https://github.com/SonarSource/sonarqube/blob/b40187fc07b4949ec4f0ff8e597edb1ff4fc8b2f/server/sonar-server/src/main/java/org/sonar/server/setting/ws/SetAction.java#L77): Remove this unused "i18n" private field. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1068)

File: sonar-plugin-api/src/main/java/org/sonar/api/config/PropertyDefinition.java [hunk 2]
```diff
@@ -155,44 +164,59 @@ public static Result validate(PropertyType type, @Nullable String value, List<St
...
+    return validations.getOrDefault(type, aValue -> Result.SUCCESS).validate(value);
   }
 
-  private static Result validateBoolean(String value) {
-    if (!StringUtils.equalsIgnoreCase(value, "true") && !StringUtils.equalsIgnoreCase(value, "false")) {
-      return Result.newError("notBoolean");
-    }
-    return Result.SUCCESS;
+  private static EnumMap<PropertyType, Validation> createValidations(List<String> options) {
+    EnumMap<PropertyType, Validation> validations = new EnumMap<>(PropertyType.class);
```
Comment (julienlancelot) [inline sonar-plugin-api/src/main/java/org/sonar/api/config/PropertyDefinition.java]: Nice to have : You could even create the validations in only one line by using :
```
validations = new EnumMap<>(ImmutableMap.of(BOOLEAN, validateBoolean(), INTEGER, validateInteger(), etc.));
```

File: server/sonar-server/src/main/java/org/sonar/server/setting/ws/SettingValidations.java [hunk 3]
```diff
@@ -103,6 +91,41 @@ public SettingData(String key, List<String> values, @Nullable ComponentDto compo
...
+      }
+
+      if (definition.type() == PropertyType.METRIC) {
+        metric(data);
+      } else {
+        otherTypes(data, definition);
+      }
+    }
 
+    private void otherTypes(SettingData data, PropertyDefinition definition) {
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/setting/ws/SettingValidations.java]: Please rename to validateOtherTypes

File: server/sonar-server/src/main/java/org/sonar/server/setting/ws/SettingValidations.java [hunk 4]
```diff
@@ -103,6 +91,41 @@ public SettingData(String key, List<String> values, @Nullable ComponentDto compo
...
+        .map(definition::validate)
+        .filter(result -> !result.isValid())
+        .findAny()
+        .ifPresent(result -> {
+          throw new BadRequestException(i18n.message(Locale.ENGLISH, "property.error." + result.getErrorKey(),
+            format("Error when validating setting with key '%s' and value [%s]", data.key, data.values.stream().collect(Collectors.joining(", ")))));
+        });
+    }
+
+    private void metric(SettingData data) {
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/setting/ws/SettingValidations.java]: Please rename to validateMetric or validateTypeMetric

File: server/sonar-server/src/main/java/org/sonar/server/setting/ws/SettingValidations.java [hunk 5]
```diff
@@ -103,6 +91,41 @@ public SettingData(String key, List<String> values, @Nullable ComponentDto compo
...
+      try (DbSession dbSession = dbClient.openSession(false)) {
+        List<MetricDto> metrics = dbClient.metricDao().selectByKeys(dbSession, data.values);
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/setting/ws/SettingValidations.java]: Please filter out disable metrics, it should not be allow to set a disable metric, whereas it's possible as it is now.

File: server/sonar-server/src/main/java/org/sonar/server/setting/ws/SettingValidations.java [hunk 6]
```diff
@@ -127,5 +130,13 @@ private void metric(SettingData data) {
           data.key, data.values.stream().collect(Collectors.joining(", ")));
       }
     }
+
+    private void login(SettingData data) {
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/setting/ws/SettingValidations.java]: Please rename to validateLogin or validateTypeLogin

File: server/sonar-server/src/main/java/org/sonar/server/setting/ws/SettingValidations.java [hunk 7]
```diff
@@ -127,5 +130,13 @@ private void metric(SettingData data) {
...
+      try (DbSession dbSession = dbClient.openSession(false)) {
+        List<UserDto> users = dbClient.userDao().selectByLogins(dbSession, data.values);
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/setting/ws/SettingValidations.java]: Please filter out disable users

**Comment by julienlancelot** [comment]
Apply my feedbacks then LGTM

**Pull request merged**